sos/plugins/juju.py

```diff
@@ -37,85 +19,59 @@ class Juju(Plugin, UbuntuPlugin):
 
     plugin_name = 'juju'
     profiles = ('virt', 'sysmgmt')
-    files = ('/usr/bin/juju', '/usr/bin/juju-run', '/snap/bin/juju')
 
-    option_list = [
-        ('export-mongodb',
-         'Export mongodb collections as json files', '', False),
-        ('generate-bundle',
-         """Generate a YAML bundle of the current environment
-         (requires juju-deployerizer)""", '', False),
-    ]
-
-    def get_deployed_services(self):
-        cmd = "juju status --format json"
-        status_json = self.call_ext_prog(cmd)['output']
-        self.add_string_as_file(status_json, "juju_status_json")
-        # if status_json isn't in JSON format (i.e. 'juju' command not found),
-        # or if it does not contain 'services' key, return empty list
-        try:
-            return json_loads(status_json)['services'].keys()
-        except ValueError:
-            return []
-
-    @ensure_service_is_running("juju-db")
-    def export_mongodb(self):
-        collections = (
-            "relations",
-            "environments",
-            "linkednetworks",
-            "system",
-            "settings",
-        )
-
-        for collection in collections:
-            self.add_cmd_output(
-                "/usr/lib/juju/bin/mongoexport --ssl \
-                --dbpath=/var/lib/juju/db --db juju --collection {0} \
-                --jsonArray".format(collection),
-                suggest_filename="{}.json".format(collection))
+    # Using files instead of packages here because there is no identifying
+    # package on a juju machine.
+    files = ('/var/log/juju')
 
     def setup(self):
-        self.add_copy_spec("/var/log/upstart/juju-db.log")
-        self.add_copy_spec("/var/log/upstart/juju-db.log.1")
-        if not self.get_option("all_logs"):
-            # We need this because we want to collect to the limit of all
-            # *.logs in the directory.
-            if(os.path.isdir("/var/log/juju/")):
-                for filename in os.listdir("/var/log/juju/"):
-                    if filename.endswith(".log"):
-                        fullname = os.path.join("/var/log/juju/" + filename)
-                        self.add_copy_spec(fullname)
-            self.add_cmd_output('ls -alRh /var/log/juju*')
-            self.add_cmd_output('ls -alRh /var/lib/juju/*')
+        # Get juju services, they are uniquely named based on the deployment
+        # so we need to parse what their names are.
+        cmd = "systemctl --no-pager --all --no-legend list-units juju\\*"
+        juju_systemctl_output = self.exec_cmd(cmd)["output"]
+        for line in iter(juju_systemctl_output.splitlines()):
+            match = re.match(r"^(juju\S+)\.", line)
+            if match:
+                self.add_journal(units=str(match.group(1)))
+                self.add_service_status(str(match.group(1)))
```

Comment:
Assuming the `juju*` services are actually well-known names, this would be made a lot cleaner/simpler by using an `InitSystem` in the Ubuntu policy. E.G.

```python
class Juju(Plugin, UbuntuPlugin):
    services = ('juju', 'other_juju', 'another_juju_service')

    def setup(self):
    for service in services:
        self.add_journal(service)
        self.add_service_status(service)
```
Note that by defining the `services` class attr, you also ensure plugin enablement when any of those services are found.

And if they aren't well known, the `SystemdInit` class could be easily extended to allow searching the discovered services (and discovering them is already handled today)

In any event, `self.add_service_status` will _not_ work without an `InitSystem` being enabled by the policy.

Reply:
+1.

The only advantage of the current implementation is in no need to maintain the list of `juju` services (that needs some coordination/awareness/effort). But the cost is uglier code, slower to execute. Due to this argument, I would prefer the Jake's approach but I am OK also with the current.

Reply:
Hi,
It seems that https://github.com/sosreport/sos/pull/1936 enables the policy by default and there is no need to initiate it on the plugin.

Unfortunately, there is no better way right now to find the correct juju-* services without this trick.

Reply:
So the service names are not well-known? Seems strange, but we can remedy this by extending `SystemdInit` to provide a method to return service names matching a regex. I'll work on something for this today.

Reply:
Opened #1943 for this. I'll merge that once we get another maintainer's ack and then this PR can be updated to use the new `get_service_names()` method.

Once that is done, this PR has my ack.

Reply:
The service names are unfortunately generated - and contain a number based on the deployment..

With your PR. this just can let us make the whole bit something like:
for service in self.get_service_names("juju*"):
    self.add_journal(service)
    self.add_service_status(service)

Thanks!

Reply:
Fixed on https://github.com/sosreport/sos/pull/1935/commits/cea2ccbbd985212b8c2c0aaea31fa14b5d267ce2
Did a test and the juju units are discovered and the journal and service status file are properly generated.